wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Create/Logic/GroupConversationCreationCoordinator.swift

```diff
@@ -0,0 +1,218 @@
+                name: name,
+                team: team,
+                allowGuests: allowGuests,
+                allowServices: allowServices,
+                readReceipts: readReceipts,
+                messageProtocol: messageProtocol
+            )
+        }
+}
+
+class GroupConversationCreationCoordinator {
+    private enum State {
+        case ready
+        case initializing
+        case initialized(eventHandler: (GroupConversationCreationEvent) -> Void)
+        case beginingToCreateConversation(eventHandler: (GroupConversationCreationEvent) -> Void)
+        case creatingConversation(conversation: ZMConversation, eventHandler: (GroupConversationCreationEvent) -> Void)
+        case finalizing
+    }
+    private var state: State = .ready
+
+    private var conversation: ZMConversation? {
+        guard case .creatingConversation(let conversation, _) = state else { return nil }
+        return conversation
+    }
+
+    private var eventHandler: ((GroupConversationCreationEvent) -> Void)? {
+        switch state {
+        case .creatingConversation(_, let completionHandler), .initialized(let completionHandler), .beginingToCreateConversation(let completionHandler):
+            return completionHandler
+        case .ready, .finalizing, .initializing:
+            return nil
+        }
+    }
+    
+    private var creator: GroupConversationCreator
+    
+    init(creator: GroupConversationCreator) {
+        self.creator = creator
+    }
+
+    func initialize(eventHandler: @escaping (GroupConversationCreationEvent) -> Void) -> Bool {
+        guard case .ready = state else { return false }
+        state = .initializing
+        NotificationCenter.default.addObserver(self, selector: #selector(missingLegalConsentErrorHandler), name: ZMConversation.missingLegalHoldConsentNotificationName, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(nonFederatingBackendsErrorHandler), name: ZMConversation.nonFederatingBackendsNotificationName, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(unsupportedErrorsHandler), name: ZMConversation.unknownResponseErrorNotificationName, object: nil)
+        NotificationCenter.default.addObserver(self, selector: #selector(conversationCreatedHandler), name: ZMConversation.insertedConversationUpdatedNotificationName, object: nil)
+        state = .initialized(eventHandler: eventHandler)
+        return true
+    }
+
+    func createConversation(
+        withUsers users: UserSet?,
+        name: String?,
+        allowGuests: Bool,
+        allowServices: Bool,
+        enableReceipts: Bool,
+        encryptionProtocol: EncryptionProtocol,
+        userSession: UserSessionInterface,
+        moc: NSManagedObjectContext
+     ) -> Bool {
+         guard case .initialized(let eventHandler) = state else { return false }
+         guard
+             let users = users
+         else {
+             return false
+         }
```

Comment:
I think this should work
```suggestion
         guard
             case .initialized(let eventHandler) = state,
             let users = users
         else {
             return false
         }
```

Also, throwing an error might be better than returning a boolean value ?